File: src/test/java/net/sourceforge/reb4j/IPAddressExampleTest.java
package net.sourceforge.reb4j;

import static org.hamcrest.CoreMatchers.is;
import static org.junit.Assert.assertThat;

import java.util.regex.Pattern;

import org.junit.Test;



/**
 * Unit test case that demonstrates the IP address example that is posted on the website.
 * 
 * @author Richard W. Eggert II
 *
 */
public class IPAddressExampleTest
{
	private final Regex oneDigitOctet = 
		CharClass.POSIX_DIGIT;
	private final Regex twoDigitOctet = 
		CharClass.range('1', '9').then(CharClass.POSIX_DIGIT);
	private final Regex oneHundredsOctet =
		Regex.literal("1").then(CharClass.POSIX_DIGIT.repeat(2));
	private final Regex lowTwoHundredsOctet =
		Regex.literal("2").then(CharClass.range('0', '4')).then(CharClass.POSIX_DIGIT);
	private final Regex highTwoHundredsOctet =
		Regex.literal("25").then(CharClass.range('0', '5'));
	private final Regex octet =
		Regex.or(
				oneDigitOctet, 
				twoDigitOctet, 
				oneHundredsOctet, 
				lowTwoHundredsOctet, 
				highTwoHundredsOctet
			);
	private final Regex dottedDecimalIPAddress =
		octet.then(Regex.literal(".")).repeat(3).then(octet);
	
	
	
	@Test
	public void testOneDigitOctet()
	{
		final Pattern pattern = oneDigitOctet.toPattern();
		for (int i = 0; i < 10; i++)
		{
			String s = Integer.toString(i);
			assertThat(pattern.matcher(s).matches(), is(true));
		}
		assertThat(pattern.matcher("10").matches(), is(false));
	}
	
	
	@Test
	public void testTwoDigitOctet()
	{
		final Pattern pattern = twoDigitOctet.toPattern();
		for (int i = 10; i < 100; i++)
		{
			String s = Integer.toString(i);
			assertThat(s, pattern.matcher(s).matches(), is(true));
		}
		assertThat("100", pattern.matcher("100").matches(), is(false));
		assertThat("9", pattern.matcher("9").matches(), is(false));
	}
	
	
	@Test
	public void testOneHundredsOctet()
	{
		final Pattern pattern = oneHundredsOctet.toPattern();
		for (int i = 100; i < 200; i++)
		{
			String s = Integer.toString(i);
			assertThat(s, pattern.matcher(s).matches(), is(true));
		}
		assertThat("200", pattern.matcher("200").matches(), is(false));
		assertThat("99", pattern.matcher("99").matches(), is(false));
	}
	
	
	@Test
	public void testLowTwoHundredsOctet()
	{
		final Pattern pattern = lowTwoHundredsOctet.toPattern();
		for (int i = 200; i < 250; i++)
		{
			String s = Integer.toString(i);
			assertThat(s, pattern.matcher(s).matches(), is(true));
		}
		assertThat("250", pattern.matcher("250").matches(), is(false));
		assertThat("199", pattern.matcher("199").matches(), is(false));
	}
	
	
	@Test
	public void testHighTwoHundredsOctet()
	{
		final Pattern pattern = highTwoHundredsOctet.toPattern();
		for (int i = 250; i < 256; i++)
		{
			String s = Integer.toString(i);
			assertThat(s, pattern.matcher(s).matches(), is(true));
		}
		assertThat("256", pattern.matcher("256").matches(), is(false));
		assertThat("249", pattern.matcher("249").matches(), is(false));
	}
	
	
	@Test
	public void testOctet()
	{
		final Pattern pattern = octet.toPattern();
		for (int i = 0; i < 256; i++)
		{
			String s = Integer.toString(i);
			assertThat(s, pattern.matcher(s).matches(), is(true));
		}
		assertThat("256", pattern.matcher("256").matches(), is(false));
		assertThat("a", pattern.matcher("a").matches(), is(false));
		assertThat("1111", pattern.matcher("1111").matches(), is(false));
	}
	
	
	
	@Test
	public void testDottedDecimalIPAddress()
	{
		final Pattern pattern =
			dottedDecimalIPAddress.toPattern();
		assertThat("1.2.3.4", pattern.matcher("1.2.3.4").matches(), is(true));
		assertThat("123.239.35.254", pattern.matcher("123.239.35.254").matches(), is(true));
		assertThat("1.256.3.4", pattern.matcher("1.256.3.4").matches(), is(false));
		assertThat("1.2.a.4", pattern.matcher("1.2.a.4").matches(), is(false));
	}
}

File: net.sourceforge.reb4j.scala/src/test/scala/net/sourceforge/reb4j/scala/test/IPAddressExampleTest.scala
package net.sourceforge.reb4j.scala.test
import org.junit.runner.RunWith
import org.scalatest.junit.JUnitRunner
import org.scalatest.FunSuite
import org.scalatest.Suite
import net.sourceforge.reb4j.scala.charclass.CharClass
import org.scalatest.matchers.ShouldMatchers
import net.sourceforge.reb4j.scala.Literal
import net.sourceforge.reb4j.scala.Group

@RunWith(classOf[JUnitRunner])
class IPAddressExampleTest extends Suite with ShouldMatchers
{
	val oneDigitOctet = CharClass.Posix.digit
	val twoDigitOctet = CharClass.range('1', '9') + CharClass.Posix.digit
	val oneHundredsOctet = Literal('1') + (CharClass.Posix.digit ## 2)
	val lowTwoHundredsOctet = Literal('2') + CharClass.range('0', '4') + CharClass.Posix.digit
	val highTwoHundredsOctet = Literal("25") + CharClass.range('0', '5')
	val octet = Group(oneDigitOctet||twoDigitOctet||oneHundredsOctet||lowTwoHundredsOctet||highTwoHundredsOctet)
	val dottedDecimalIPAddress = octet + (Group(Literal('.') + octet) ## 2)
	
	def testOneDigitOctet()
	{
		val pattern = oneDigitOctet.toPattern
		for (i <- 0 to 9)
			pattern.matcher(i.toString()).matches() should be (true)
		for (i <- 10 to 1000)
			pattern.matcher(i.toString()).matches() should be (false)
	}
	
	def testTwoDigitOctet()
	{
		val pattern = twoDigitOctet.toPattern
		for (i <- 0 to 9)
			pattern.matcher(i.toString()).matches() should be (false)
		for (i <- 10 to 99)
			pattern.matcher(i.toString()).matches() should be (true)
		for (i <- 100 to 1000)
			pattern.matcher(i.toString()).matches() should be (false)
	}
	
	def testOneHundredsOctet()
	{
		val pattern = oneHundredsOctet.toPattern
		for (i <- 0 to 99)
			pattern.matcher(i.toString()).matches() should be (false)
		for (i <- 100 to 199)
			pattern.matcher(i.toString()).matches() should be (true)
		for (i <- 200 to 1000)
			pattern.matcher(i.toString()).matches() should be (false)
	}
	
	def testLowTwoHundredsOctet()
	{
		val pattern = lowTwoHundredsOctet.toPattern
		for (i <- 0 to 199)
			pattern.matcher(i.toString()).matches() should be (false)
		for (i <- 200 to 249)
			pattern.matcher(i.toString()).matches() should be (true)
		for (i <- 250 to 1000)
			pattern.matcher(i.toString()).matches() should be (false)
	}
	
	def testHighTwoHundredsOctet()
	{
		val pattern = highTwoHundredsOctet.toPattern
		for (i <- 0 to 249)
			pattern.matcher(i.toString()).matches() should be (false)
		for (i <- 250 to 255)
			pattern.matcher(i.toString()).matches() should be (true)
		for (i <- 256 to 1000)
			pattern.matcher(i.toString()).matches() should be (false)
	}
	
	def testOctet()
	{
		val pattern = octet.toPattern
		for (i <- 0 to 255)
			pattern.matcher(i.toString()).matches() should be (true)
		for (i <- 256 to 1000)
			pattern.matcher(i.toString()).matches() should be (false)
		for (i <- 'a' to 'z')
			pattern.matcher(i.toString()).matches() should be (false)
	}
	
	
	def testDottedDecimalIPAddress()
	{
		val pattern = dottedDecimalIPAddress.toPattern
		pattern.matcher("1.2.3.4").matches() should be (true)
		pattern.matcher("123.239.35.254").matches() should be (true)
		pattern.matcher("1.256.3.4").matches() should be (false)
		pattern.matcher("1.2.a.4").matches() should be (false)
	}
	
	
}